doc/man/66-update.1.scd
66-scandir(1)

# NAME

66-update - Handle a format transition update

# SYNOPSYS

66-update [ *-h* ] [*-z* ] [ *-v* _verbosity_ ] [ *-l* _live_ ] [ *-d* ]  _tree_

# DESCRIPTION

*66-update* is a program that should be used  rarely. At upgrade/downgrade 
time, the inner file format used to retrieve informations of tree,service 
database currently in use, can differ between two versions. In this case, 
service databases and trees need to be rebuild to match the new format. 
The 66-update program makes a complete transition of trees and the live 
directory using a old 66 format (the one being replaced) with the new 66 format.

The goal is to ensure a smooth transition between different versions of 66 
without needing a reboot. 

# OPTIONS

*-h*
	Prints this help.

*-z*
	use color

*-v* _verbosity_
	Increases/decreases the verbosity of the command.++
	*1* : (Default) Only prints error messages.++
	*2* : Also prints warning messages.++
	*3* : Also prints tracing messages.++
	*4* : Also prints debugging messages.

*-l* _live_
	An absolute path. Create the scandir at _live_. Default is *%%livedir%%*.
	The default can also be changed at compile-time by passing the
	\--livedir=_live_ option to *./configure*. Should be under a writable
	filesystem - likely a RAM filesystem.

*-d*
	dry run. Performs operation without modify the current state of the 
	system. All command's output will be prefixed by "dry run do: " 
	sentence. It's a good idea to first use this option before performing 
	a real update.

If no _tree_ is given, all trees of the owner (root or user) of the 
process will be processed. In the case of user owned trees *66-update*
must run separately for each user.

# TRANSITION PROCESS

It determines if the tree is already initialized. In such case, it
makes a copy of the current database on a temporary directory and
switches the database to it.

It saves the current state of the tree (current,enabled,...).

It destroys the tree with *66-tree -R* command. 

It creates the tree with *66-tree -n(cEa)* command, reproducing the options
of the saved state.

It enables the service(s) previously declared on the _tree_.

It switches the current database to the newly created database by the enable process.

At the end of the process, the tree and the live state has not changed. Services already is use
are not stopped or restarted during the transition process and match exactly the same state
before and after the process.

It tries to be safe as possible and exit 111 is a problem occurs. In any case 
on upgrade/downgrade time, 66 tools will not work properly if the transition 
is not made smoothly even by the sysadmin itself. So, we prefer to crash instead of
leaving an inconsistent state and let the sysadmin correct the things.


NEWS
Changelog for 66

In 0.2.5.2
----------

    - Fix build: Remove bytes.h oblibs header file
    
In 0.2.5.1
----------
    
    - Bugs fix
    - 66-tree -S options: if after_tree and tree have the same name,
    tree is considered as the very first tree to start.
    - 66-unmountall: do not umount SS_LIVE
    - Add SIGPWR control file at creation of .s6-svscan directory

In 0.2.5.0
----------
    
    - Bugs Fix: Bad memory allocation
    - 66-tree: add -S option which allow to order the start tree process
    even after it creation.
    
In 0.2.4.1
----------
    
    - Hot fix: fix html documentation

In 0.2.4.0
----------
    
    - Bugs fix: Bad memory allocation.
    
    - Enable again RB_DISABLE_CAD but don't crash if its not available
    and warn user
    
    - Add @optsdepends and @extdepends field at [main] section:
        - @optsdepends can be considere as: "enable one on this service 
        or none"
        - @extdepends can be considere as: "enable the service if it is 
        not already declared on a tree"
    
    - 66-in{tree,service}: 
        - in case of empty value the tools return None
        - add -n option: this avoids to display the name of the field.
        - add field at 66-intree:
            - start: displays the list of tree(s) started before
            - allowed: displays a list of allowed user to use the tree
            - symlinks: displays the target of tree's symlinks
        - add field at 66-inservice:
            - optsdepends: displays the service's optional dependencies
            - extdepends: displays the service's external dependencies
    
    - 66-shutdown skeleton: be safier and check if options are past
    
    - New tool:
        - 66-update: The 66-update program makes a complete transition 
        of trees and the live directory using a old 66 format 
        (the one being replaced) with the new 66 format.

In 0.2.3.2
----------
    
    - Fix bad annoucement at 66-svctl
    - Warn in case of bad key at parenthesis parsing process
    
In 0.2.3.1
----------
    
    - Bugs fix: bad allocation memory
    - Makefile: remove creation of empty directories. 66-tree will check
    all directories and create it if it missing.
    
In 0.2.3.0
----------
    
    - Adapt to oblibs 0.0.4.0
    - Remove RB_DISABLE_CAD at 66-boot by default and allow to enable
    it with -c option. Starting on virtual system like lxc crash with
    this option. Well be safe on every case.
    - Use of new function log_? familly from oblibs:
        - standardization of the exit code
        - standardization of the verbosity output
        - allow verbosity to 4 to display debug message
    - Doc typo fix        
    
In 0.2.2.2
----------
    
    - Adapt to skalibs 2.9.1.0
    - Remove insta_? deprecated function
    
In 0.2.2.1
----------
    
    - Typo fix at html documentation
    - Fix 66-intree without options
    
In 0.2.2.0
----------

    - Bugs fix: Always check the existing of the 66 heart directories
    - 66-info is now deprecated and splitted into two differents API.
    66-intree give informations about tree
    66-inservice give informations about services

In 0.2.1.2
----------

    - Hot fix: fix 66-info graph display.
    
In 0.2.1.1
----------

    - Hot fix: fix the build of the service dependencies graph.

In 0.2.1.0
----------

    - Bugs fix.
   
    - A synchronization is now made on reboot even with force option.
   
    - The field @name has no effects except for instantiated service but 
    can be omitted. The name of the frontend file is took as name. 
    In case of instantiated service, the field @name must contain the 
    complete name of the frontend service file -- Refer to the frontend 
    documentation page.
    
    - New skeleton file: rc.shutdown.final. This skeleton file will be run
    at the very end of the shutdown procedure, after all processes have 
    been killed and all filesystems have been unmounted, just before the 
    system is rebooted or the power turned off.
    This script normally remains empty.
   
    - Extra tools has removed and provided as an independent programs at
    https://framagit.org/Obarun/66-tools.git except for 66-echo and 
    66-unmountall program which are a dependent part of 66.
   
    - The account to run the s6-log program at the associate service logger
    can be set at compilation time by the --with-s6-log-user=USER options.
    The default is root. Obviously, the @runas field at the [logger] section
    overwritte it.
    This options set the account to run the uncaught-logs too. Also,
    this option can be overwritten with the '-l' option at 66-boot invocation
    and the '-L' option at 66-scandir invocation. 
    
In 0.2.0.4
----------
	
	- Bugs fix on memory allocation
	- Respect policies decision of user: 
		Remove -m option from 66-boot on init file
	- Add -d feature to 66-scandir 

In 0.2.0.3
----------

	- 66-parser: Fix write of configuration file, add -c|C features 
	
In 0.2.0.2
----------

	- Fix the written of user configuration file 
	- Fix oneshot and bundle status check
	- Add -l features to 66-tree, rewrite tree_unsupervise function
	
In 0.2.0.0
----------

	- New tools:
		- 66-boot.
		- 66-scanctl.
		- 66-shutdown.
		- 66-shutdownd.
		- 66-hpr.
		- 66-env.
		- 66-parser.
		- 66-which.
		- 66-echo.
		- 66-unmountall.
		- execl-subuidgid.
	- New @key field and change on frontend file:
		- [Logger] section accept a @depends field, @timestamp accept none
		as value, readiness notification is a default.
		- @hiercopy in [main] allow to copy any file or directory coming
		from the directory of the service.
	- 66-envfile is now deprecated, use execl-envfile in replacement.
	- 66-enable: add -F, -c, -C features.
	- 66-svctl is now asynchrone to launch services.
	- Man pages are available.
	- Environment variables limitation: maximum 100 files by directory,
	each file cannot contain more than 4096 bytes or 50 variables.

src/66/66-update.c
/* 
 * 66-update.c
 * 
 * Copyright (c) 2018-2020 Eric Vidal <eric@obarun.org>
 * 
 * All rights reserved.
 * 
 * This file is part of Obarun. It is subject to the license terms in
 * the LICENSE file found in the top-level directory of this
 * distribution.
 * This file may not be copied, modified, propagated, or distributed
 * except according to the terms contained in the LICENSE file./
 */

#include <stdio.h>
#include <unistd.h>
#include <string.h>

#include <oblibs/log.h>
#include <oblibs/string.h>
#include <oblibs/sastr.h>
#include <oblibs/types.h>
#include <oblibs/directory.h>

#include <skalibs/stralloc.h>
#include <skalibs/buffer.h>
#include <skalibs/sgetopt.h>
#include <skalibs/djbunix.h>
#include <skalibs/unix-transactional.h>

#include <66/ssexec.h>
#include <66/constants.h>
#include <66/utils.h>
#include <66/db.h>
#include <66/svc.h>
#include <66/tree.h>
#include <66/backup.h>
#include <66/resolve.h>
#include <66/parser.h>

#define USAGE "66-update [ -h ] [ -z ] [ -v verbosity ] [ -l live ] [ -d ] tree(s)"

static stralloc WORKDIR = STRALLOC_ZERO ;
static uint8_t DRYRUN = 0 ;
static char *drun = "dry run do: " ;
static inline void info_help (void)
{
  static char const *help =
"66-update <options> tree(s)\n"
"\n"
"options :\n"
"	-h: print this help\n"
"	-z: use color\n"
"	-v: increase/decrease verbosity\n"
"	-l: live directory\n"
"	-d: dry run\n"
"\n"
"if no tree is given, all trees will be processed.\n"
;

 if (buffer_putsflush(buffer_1, help) < 0)
    log_dieusys(LOG_EXIT_SYS, "write to stdout") ;
}

static void cleanup(void)
{
	int e = errno ;
	if (WORKDIR.len)
	{
		log_trace(DRYRUN ? drun : "","delete temporary directory: ",WORKDIR.s) ;
		rm_rf(WORKDIR.s) ;
	}
	errno = e ;
}

int tree_is_current(char const *base,char const *treename,uid_t owner)
{
	stralloc sacurr = STRALLOC_ZERO ;
	int current = 0 ;
	
	if (tree_find_current(&sacurr,base,owner))
	{
		char name[sacurr.len + 1] ;
		if (!ob_basename(name,sacurr.s)) log_dieu_nclean(LOG_EXIT_SYS,&cleanup,"basename of: ",sacurr.s) ;
		current = obstr_equal(treename,name) ;
	}
	stralloc_free(&sacurr) ;
	return current ;
}

int tree_is_enabled(char const *treename)
{
	return tree_cmd_state(VERBOSITY,"-s",treename) ;	
}

void tree_allowed(stralloc *list,char const *base, char const *treename)
{
	stralloc sa = STRALLOC_ZERO ;
	size_t treenamelen = strlen(treename), baselen = strlen(base), pos ;
	char tmp[baselen + SS_SYSTEM_LEN + 1 + treenamelen + SS_RULES_LEN + 1] ;
	
	auto_strings(tmp,base,SS_SYSTEM,"/",treename,SS_RULES) ;
	
	if (!sastr_dir_get(&sa,tmp,"",S_IFREG))
		log_dieusys_nclean(LOG_EXIT_SYS,&cleanup,"get permissions of tree at: ",tmp) ;

	for (pos = 0 ;pos < sa.len; pos += strlen(sa.s + pos) + 1)
	{
		char *suid = sa.s + pos ;
		uid_t uid = 0 ;
		if (!uid0_scan(suid, &uid))
			log_dieusys_nclean(LOG_EXIT_SYS,&cleanup,"get uid of: ",suid) ;
		if (pos) 
			if (!stralloc_cats(list,",")) log_die_nomem_nclean(&cleanup,"stralloc") ;
		if (!get_namebyuid(uid,list))
			log_dieusys_nclean(LOG_EXIT_SYS,&cleanup,"get name of uid: ",suid) ;
	}
	if (!stralloc_0(list)) log_die_nomem_nclean(&cleanup,"stralloc") ;
	log_trace(DRYRUN ? drun : "","allowed user(s) for tree: ",treename," are: ",list->s) ;
	stralloc_free(&sa) ;
}

void tree_contents(stralloc *list,char const *tree,ssexec_t *info)
{
	stralloc sa = STRALLOC_ZERO ;
	size_t treelen = strlen(tree), pos, newlen ;
	char solve[treelen + SS_SVDIRS_LEN + SS_DB_LEN + SS_SRC_LEN + 1] ;
	newlen = treelen + SS_SVDIRS_LEN ;
	auto_strings(solve,tree,SS_SVDIRS,SS_SVC) ;
	
	if (!sastr_dir_get(&sa,solve,"",S_IFDIR))
		log_dieusys_nclean(LOG_EXIT_SYS,&cleanup,"get source svc service file of: ",tree) ;

	auto_string_from(solve,newlen,SS_DB,SS_SRC) ;

	if (!sastr_dir_get(&sa,solve,SS_MASTER + 1,S_IFDIR))
		log_dieusys_nclean(LOG_EXIT_SYS,&cleanup,"get source of atomic service file of: ",tree) ;
	
	for (pos = 0 ;pos < sa.len; pos += strlen(sa.s + pos) + 1)
	{
		char *name = sa.s + pos ;
		int logname = get_rstrlen_until(name,SS_LOG_SUFFIX) ;
		if (logname > 0) continue ;
		log_trace(DRYRUN ? drun : "","tree: ",info->treename.s," contain service: ",name) ;
		if (ss_resolve_src_path(list,name,info->owner) < 1)
			log_dieu_nclean(LOG_EXIT_SYS,&cleanup,"resolve source path of: ",name) ;
	
	}
	stralloc_free(&sa) ;
}

static int run_cmdline(char const *prog,char const **add,int len,char const *const *envp)
{
	pid_t pid ;
	int wstat ;
	
	char fmt[UINT_FMT] ;
	fmt[uint_fmt(fmt, VERBOSITY)] = 0 ;
	
	int m = 4 + len, i = 0, n = 0 ;
	char const *newargv[m] ;
	newargv[n++] = prog ;
	newargv[n++] = "-v" ; 
	newargv[n++] = fmt ;
	for (;i<len;i++)
		newargv[n++] = add[i] ;
	newargv[n] = 0 ;
	
	pid = child_spawn0(newargv[0],newargv,envp) ; 
	if (waitpid_nointr(pid,&wstat, 0) < 0)
		log_dieusys_nclean(LOG_EXIT_SYS,&cleanup,"wait for: ",newargv[0]) ;
	if (wstat) return 0 ;
	
	return 1 ;
}


int main(int argc, char const *const *argv,char const *const *envp)
{
	int r ;
	unsigned int nclassic = 0, nlongrun = 0, nbsv = 0 ;
	size_t systemlen, optslen = 5, pos, len ;
	char tree_opts_create[optslen] ;
	char *fdir = 0 ;

	log_color = &log_color_disable ;

	stralloc satree = STRALLOC_ZERO ;
	stralloc allow = STRALLOC_ZERO ;
	stralloc contents = STRALLOC_ZERO ;
	ssexec_t info = SSEXEC_ZERO ;

	PROG = "66-update" ;	
	{
		subgetopt_t l = SUBGETOPT_ZERO ;
		for (;;)
		{
			int opt = subgetopt_r(argc,argv, "hzv:l:dc", &l) ;
			
			if (opt == -1) break ;
			switch (opt)
			{
				case 'h' : 	info_help(); return 0 ;
				case 'v' :  if (!uint0_scan(l.arg, &VERBOSITY)) log_usage(USAGE) ; break ;
				case 'l' : 	if (!stralloc_cats(&info.live,l.arg)) log_die_nomem("stralloc") ;
							if (!stralloc_0(&info.live)) log_die_nomem("stralloc") ;
							break ;
				case 'z' :	log_color = !isatty(1) ? &log_color_disable : &log_color_enable ; break ;
				case 'd' :  DRYRUN = 1 ; break ;
				case 'c' :	log_die(LOG_EXIT_SYS,"deprecated option -- please use -z instead") ;
				default	:	log_usage(USAGE) ; 
			}
		}
		argc -= l.ind ; argv += l.ind ;
	}
	
	info.owner = getuid() ;
	
	if (!set_ownersysdir(&info.base,info.owner)) log_dieusys(LOG_EXIT_SYS, "set owner directory") ;

	char system[info.base.len + SS_SYSTEM_LEN + 2] ;
	auto_strings(system,info.base.s,SS_SYSTEM,"/") ;
	systemlen = info.base.len + SS_SYSTEM_LEN + 1 ;
	
	if (!argc)
	{
		if (!sastr_dir_get(&satree,system,SS_BACKUP + 1,S_IFDIR))
			log_dieusys(LOG_EXIT_SYS,"get list of trees at: ",system) ;

		if (!satree.len)
		{
			log_info(DRYRUN ? drun : "","No trees exist yet -- Nothing to do") ;
			goto exit ;
		}
	}
	else
	{
		int i = 0 ;
		size_t arglen = 0 ;
		for (;i < argc ; i++)
		{
			arglen = strlen(argv[i]) ;
			char tree[systemlen + arglen + 1] ;
			auto_strings(tree,system,argv[i]) ;
			r = scan_mode(tree,S_IFDIR) ;
			if (r == -1) { errno = EEXIST ; log_diesys(LOG_EXIT_SYS,"conflicting format of: ",tree) ; }
			if (!r) log_die(LOG_EXIT_USER,"tree: ",tree," doesn't exist") ;
			if (!stralloc_catb(&satree,argv[i],strlen(argv[i]) + 1))
				log_die_nomem("stralloc") ;
		}
	}
	
	len = satree.len ;
	for (pos = 0 ; pos < len; pos += strlen(satree.s + pos) + 1)
	{
		int dbok = 0 ;
		
		nclassic = nlongrun = nbsv = 0 ;
		info.base.len = info.tree.len = info.treename.len = 0 ;
		allow.len = WORKDIR.len = contents.len = 0 ;
		auto_strings(tree_opts_create,"-n") ;
		optslen = 2 ;
		
		if (!auto_stra(&info.tree,satree.s + pos))
			log_die_nomem("stralloc") ;
 		
		set_ssinfo(&info) ;
		
		char tmp[systemlen + SS_BACKUP_LEN + info.treename.len + SS_SVDIRS_LEN + SS_DB_LEN + 1] ;
		char current[info.livetree.len + 1 + info.treename.len + 9 + 1] ;
		
		log_info(DRYRUN ? drun : "","save state of tree: ", info.treename.s) ;
		if (tree_is_current(info.base.s,info.treename.s,info.owner))
		{
			log_trace(DRYRUN ? drun : "","tree: ",info.treename.s," is marked current") ;
			auto_string_from(tree_opts_create,optslen,"c") ;
			optslen +=1 ;
		}
		if (tree_is_enabled(info.treename.s) == 1)
		{
			log_trace(DRYRUN ? drun : "","tree: ",info.treename.s," is marked enabled") ;
			auto_string_from(tree_opts_create,optslen,"E") ;
			optslen +=1 ;
		}
		
		tree_allowed(&allow,info.base.s,info.treename.s) ;
		
		log_info(DRYRUN ? drun : "","save service(s) list of tree: ", info.treename.s) ;
		tree_contents(&contents,info.tree.s,&info) ;
		
		dbok = db_ok(info.livetree.s, info.treename.s) ;
		
		if (dbok)
		{
			fdir = 0 ;
			
			log_trace(DRYRUN ? drun : "","find current source of live db: ",info.livetree.s,"/",info.treename.s) ;
			r = db_find_compiled_state(info.livetree.s,info.treename.s) ;
			if (r == -1) log_die(LOG_EXIT_SYS,"inconsistent state of: ",info.livetree.s) ;
			if (r == 1) {
				auto_strings(tmp,system,SS_BACKUP + 1,"/",info.treename.s,SS_DB) ;
			}
			else auto_strings(tmp,system,info.treename.s,SS_SVDIRS,SS_DB) ;

			log_trace(DRYRUN ? drun : "","created temporary directory at: /tmp") ;
			fdir = dir_create_tmp(&WORKDIR,"/tmp",info.treename.s) ;
			if (!fdir) log_dieusys_nclean(LOG_EXIT_SYS,&cleanup,"create temporary directory") ;
			
			log_trace(DRYRUN ? drun : "","copy contents of: ",tmp," to: ",WORKDIR.s) ;
			if (!hiercopy(tmp,fdir)) log_dieusys_nclean(LOG_EXIT_SYS,&cleanup,"copy: ",tmp," to: ",WORKDIR.s) ;
			
			char new[WORKDIR.len + 1 + info.treename.len + 1] ;
					
			auto_strings(current,info.livetree.s,"/",info.treename.s,"/compiled") ;
			auto_strings(new,WORKDIR.s,"/",info.treename.s) ;
			
			log_info(DRYRUN ? drun : "","update ",info.livetree.s,"/",info.treename.s," to: ",new) ;
			
			if (!DRYRUN)
				if (!atomic_symlink(new, current, PROG))
					log_dieusys(LOG_EXIT_SYS,"update: ",current," to: ", new) ;
		
			if (DRYRUN) cleanup() ;
		}
		
		/** finally we can destroy the tree and recreate it*/
		// destroy
		{
			char const *t[] = { "-l", info.live.s, "-R", info.treename.s } ;
			if (DRYRUN)
			{
				log_info(drun,"66-tree -l ",info.live.s," -R ",info.treename.s) ; 
			}
			else
			{
				if (!run_cmdline(SS_EXTBINPREFIX "66-tree",t,4,envp))
					log_dieu(LOG_EXIT_SYS,"delete tree: ", info.treename.s) ;
			}
		}
		//create
		{
			char const *t[] = { "-l", info.live.s, tree_opts_create,"-a",allow.s, info.treename.s } ;
			if (DRYRUN)
			{
				log_info(drun,"66-tree -l ",info.live.s," ",tree_opts_create," -a ",allow.s," ",info.treename.s) ; 
			}
			else
			{
				if (!run_cmdline(SS_EXTBINPREFIX "66-tree",t,6,envp))
					log_dieu(LOG_EXIT_SYS,"create tree: ", info.treename.s) ;
			}
		}
		/* we must reimplement the enable process instead of
		 * using directly 66-enable. The 66-enable program will use
		 * is own workdir with an empty tree. At call of db_update(),
		 * the service already running will be brought down. We don't 
		 * want this behavior. A nope update is necessary.
		 * So remake the things with the enable API */
		
		if (contents.len)
		{
			stralloc tostart = STRALLOC_ZERO ;
			log_info(DRYRUN ? drun : "","enable service(s) of tree: ",info.treename.s) ;
			if (!DRYRUN)
			{
				auto_strings(tmp,system,info.treename.s,SS_SVDIRS) ;
			}
			else
			{
				WORKDIR.len = 0 ;
				log_trace(drun,"copy: ", info.tree.s,SS_SVDIRS," to a temporary directory") ;
				if (!tree_copy(&WORKDIR,info.tree.s,info.treename.s)) 
					log_dieusys_nclean(LOG_EXIT_SYS,&cleanup,"create tmp working directory") ;
				/* we need to remove the contain of .resolve directory
				 * The write process will try to read it and obviously it
				 * get a wrong information*/
				auto_strings(tmp,WORKDIR.s,SS_RESOLVE) ;
				log_trace(drun,"remove directory: ", tmp) ;
				if (rm_rf(tmp) < 0) log_dieusys_nclean(LOG_EXIT_SYS,&cleanup,"remove: ",tmp) ;
				log_trace(drun,"create directory: ", tmp) ;
				if (!dir_create_parent(tmp,0755))
					log_dieusys_nclean(LOG_EXIT_SYS,&cleanup,"create directory: ",tmp) ;
				auto_string_from(tmp,0,WORKDIR.s) ;
				
			}

			start_parser(&contents,&info,&nbsv,1) ;
			start_write(&tostart,&nclassic,&nlongrun,tmp,&gasv,&info,1,0) ;
			/** we don't care about nclassic. Classic service are copies
			 * of original and we retrieve the original at the end of the
			 * process*/
			if(nlongrun)
			{
				ss_resolve_graph_t graph = RESOLVE_GRAPH_ZERO ;
				r = ss_resolve_graph_src(&graph,tmp,0,1) ;
				if (!r)
					log_dieu(LOG_EXIT_SYS,"resolve source of graph for tree: ",info.treename.s) ;
				
				r = ss_resolve_graph_publish(&graph,0) ;
				if (r <= 0) 
				{
					if (r < 0) log_die(LOG_EXIT_USER,"cyclic graph detected") ;
					log_dieusys(LOG_EXIT_SYS,"publish service graph") ;
				}
				
				if (!ss_resolve_write_master(&info,&graph,tmp,0))
					log_dieusys(LOG_EXIT_SYS,"update inner bundle") ;
				
				ss_resolve_graph_free(&graph) ;
				if (!db_compile(tmp,info.tree.s,info.treename.s,envp))
					log_dieu(LOG_EXIT_SYS,"compile ",tmp,"/",info.treename.s) ;
			}
			stralloc_free(&tostart) ;
			freed_parser() ;
			genalloc_free(sv_alltype,&gasv) ;
			
		}
		
		if (dbok)
		{
			auto_strings(tmp,system,info.treename.s,SS_SVDIRS,SS_DB) ;
			log_info(DRYRUN ? drun : "","update db: ",info.livetree.s,"/",info.treename.s, " to: ",tmp,"/",info.treename.s) ;
			/* Be paranoid here and use db_update instead of atomic_symlink.
			 * db_update() allow to make a running test and to see
			 * if the db match exactly the same state.
			 * We prefer to die in this case instead of leaving an
			 * inconsistent state. */
			if (!DRYRUN)
				if (!db_update(tmp,&info,envp))
					log_dieu(LOG_EXIT_SYS,"update: ",info.livetree.s,"/",info.treename.s," to: ", tmp,"/",info.treename.s) ;
		}

		cleanup() ;	
		
		log_info(DRYRUN ? drun : "","tree: ",info.treename.s," updated successfully") ;
	}
	exit:
	stralloc_free(&satree) ;
	stralloc_free(&allow) ;
	stralloc_free(&WORKDIR) ;
	stralloc_free(&contents) ;
	ssexec_free(&info) ;
	return 0 ;
	
}
	
